### ISSUE TITLE
bug: cp.asarray out of memory [5.0.0b4]

### ISSUE THREAD

**Issue by mingwugmail** [cupy/cupy#1576]
```
import numpy as np
import cupy as cp

# c = cp.zeros((3001000, 600), dtype=np.float32)  # no bug if alloc directly

n = np.zeros((3001000, 600), dtype=np.float32)
c = cp.asarray(n)
```
cupy.cuda.runtime.CUDARuntimeError: cudaErrorMemoryAllocation: out of memory

cupy                      5.0.0b4                   <pip>

(BTW, same code no problem with previous version v4.x)

Ubuntu 16.04.5 LTS

cuda-9.1

```
$ nvidia-smi 
Mon Aug 27 19:37:43 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 396.44                 Driver Version: 396.44                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TIT...  Off  | 00000000:01:00.0  On |                  N/A |
| 22%   40C    P8    16W / 250W |    100MiB / 12205MiB |      1%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX TIT...  Off  | 00000000:02:00.0 Off |                  N/A |
| 22%   29C    P8    13W / 250W |      1MiB / 12212MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      2407      G   /usr/lib/xorg/Xorg                            98MiB |
+-----------------------------------------------------------------------------+
```

**Comment by mingwugmail** [comment]
maybe np.float32 is missed in cp.asarray call (become np.float64)?

**Comment by kmaehashi** [comment]
> cupy.cuda.runtime.CUDARuntimeError: cudaErrorMemoryAllocation: out of memory

Could you share the full stack trace?

**Comment by mingwugmail** [comment]
$ python3 cupybug.py 
Traceback (most recent call last):
  File "cupy/cuda/pinned_memory.pyx", line 303, in cupy.cuda.pinned_memory.PinnedMemoryPool.malloc
  File "cupy/cuda/pinned_memory.pyx", line 177, in cupy.cuda.pinned_memory._malloc
  File "cupy/cuda/pinned_memory.pyx", line 178, in cupy.cuda.pinned_memory._malloc
  File "cupy/cuda/pinned_memory.pyx", line 29, in cupy.cuda.pinned_memory.PinnedMemory.__init__
  File "cupy/cuda/runtime.pyx", line 228, in cupy.cuda.runtime.hostAlloc
  File "cupy/cuda/runtime.pyx", line 134, in cupy.cuda.runtime.check_status
cupy.cuda.runtime.CUDARuntimeError: cudaErrorMemoryAllocation: out of memory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "cupybug.py", line 8, in <module>
    c = cp.asarray(n)
  File "/miniconda3/lib/python3.6/site-packages/cupy/creation/from_data.py", line 60, in asarray
    return core.array(a, dtype, False)
  File "cupy/core/core.pyx", line 2174, in cupy.core.core.array
  File "cupy/core/core.pyx", line 2213, in cupy.core.core.array
  File "cupy/cuda/pinned_memory.pyx", line 212, in cupy.cuda.pinned_memory.alloc_pinned_memory
  File "cupy/cuda/pinned_memory.pyx", line 286, in cupy.cuda.pinned_memory.PinnedMemoryPool.malloc
  File "cupy/cuda/pinned_memory.pyx", line 308, in cupy.cuda.pinned_memory.PinnedMemoryPool.malloc
  File "cupy/cuda/pinned_memory.pyx", line 177, in cupy.cuda.pinned_memory._malloc
  File "cupy/cuda/pinned_memory.pyx", line 178, in cupy.cuda.pinned_memory._malloc
  File "cupy/cuda/pinned_memory.pyx", line 29, in cupy.cuda.pinned_memory.PinnedMemory.__init__
  File "cupy/cuda/runtime.pyx", line 228, in cupy.cuda.runtime.hostAlloc
  File "cupy/cuda/runtime.pyx", line 134, in cupy.cuda.runtime.check_status
cupy.cuda.runtime.CUDARuntimeError: cudaErrorMemoryAllocation: out of memory

**Comment by kmaehashi** [comment]
Thanks for the details.

This out of memory error is coming from pinned memory allocator (i.e. your **host** memory is not enough). In v4.2.0 and v5.0.0b2, we changed the allocation strategy of pinned memory; the allocation size is now round-up. Prior to these versions, your code requires 3001000*600*4 (approx. 7.2 GB) host memory, whereas 8.0 GB in and after these versions.

We originally thought this change does not affect most of users, but you are affected ;)
After discussion with @okuta, we decided to implement a workaround for this case.  Thanks again for reporting this.
